### ISSUE TITLE
schedule job with when and splay, but splay not factored into next_run_time

### ISSUE THREAD

**Issue by sfozz** [saltstack/salt#51374]
### Description of Issue/Question
Shedule a job on a minion using when and splay, but the splay isn't taken into account and the job still runs on the 'when' time not 'when + splay'

### Setup
(Please provide relevant configs and/or SLS files (Be sure to remove sensitive info).)
2 Debian 9 Nodes - basic install (one with salt-master & salt-minion, the other just salt-minion)
Install saltstack 2018.3.3, point minion at the master and accept the keys.

### Steps to Reproduce Issue
(Include debug logs if possible and relevant.)

Create event on minion 
`salt minion-001 schedule.add once_splay function='cmd.run' job_args="['date >> /tmp/date.log']" when="10:30am" splay=$(( 60 * 60 * 3 ))`

```
[DEBUG   ] minion return: {u'fun_args': [u'once_splay', {u'function': u'cmd.run', u'splay': 10800, u'job_args': [u'date >> /tmp/date.log'], u'when': u'10:30am'}], u'jid': u'20190126102845100505', u'retu
rn': {u'comment': u'Added job: once_splay to schedule.', u'result': True}, u'retcode': 0, u'success': True, u'fun': u'schedule.add'}                                                                      
[DEBUG   ] schedule.handle_func: Adding splay of 9322 seconds to next run.                                                                                                                                
```
```
root@minion-001:~# tail -1f /tmp/date.log                                                                                                                                                    Sat Jan 26 10:30:00 NZDT 2019
```
Which suggests that although the minion is configured with a value for splay that it isn't being used... It looks to me like the 'when' is being checked to see if it matches the current time or that the splay hasn't been added to the 'when' to offset it. So based on the minion debug log I would expect this job to run around 12:30pm.

### Versions Report
(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)

Master:
root@salt-001:~# salt --versions-report
Salt Version:
           Salt: 2018.3.3

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.5.3
      docker-py: Not Installed
          gitdb: 2.0.0
      gitpython: 2.1.1
          ioflo: Not Installed
         Jinja2: 2.9.4
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 3.5.3 (default, Sep 27 2018, 17:25:39)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: 2.0.1
        timelib: Not Installed
        Tornado: 4.4.3
            ZMQ: 4.2.1

System Versions:
           dist: debian 9.6
         locale: UTF-8
        machine: x86_64
        release: 4.9.0-8-amd64
         system: Linux
        version: debian 9.6

Minion:
Salt Version:
           Salt: 2018.3.3

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.5.3
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.9.4
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 3.5.3 (default, Sep 27 2018, 17:25:39)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.4.3
            ZMQ: 4.2.1

System Versions:
           dist: debian 9.6
         locale: UTF-8
        machine: x86_64
        release: 4.9.0-8-amd64
         system: Linux
        version: debian 9.6

**Comment by garethgreenaway** [comment]
@sfozz Thanks for the report.  `splay` should definitely be usable with `when`, we'll look into this.

**Comment by sfozz** [comment]
@garethgreenaway no worries, it looks like the problem is here

https://github.com/saltstack/salt/blob/2018.3/salt/utils/schedule.py#L1221-L1224

we are checking `_next_fire_time` rather than `_next_scheduled_fire_time`, the latter includes the value from `_splay()`

**Comment by sfozz** [comment]
Just by way of a test I updated:

https://github.com/saltstack/salt/blob/2018.3/salt/utils/schedule.py#L1222

to be:
```python
if data['_splay'] <= now <= (data['_splay'] + loop_interval)
```
And this gives the desired behaviour, caveat when a job doesn't have splay it will break. 

Here is the debug log showing that the `_next_scheduled_fire_time` is being updated to be the same as the `_splay` time.

```
[DEBUG   ] _splay: 2019-01-30 11:54:02 _next_scheduled_fire_time: 2019-01-30 11:54:00 _next_fire_time: 2019-01-30 11:54:00                     
[DEBUG   ] {'_next_fire_time': datetime.datetime(2019, 1, 30, 11, 54), '_run': True, 'function': 'cmd.run', 'splay': 10, 'args': ['date >> /tmp/splay_test.log'], 'name': 'splay_test', 'when': ['11:54am'], 'enabled': True, 'jid_include': T
rue, '_next_scheduled_fire_time': datetime.datetime(2019, 1, 30, 11, 54), '_splay': datetime.datetime(2019, 1, 30, 11, 54, 2), 'maxrunning': 1}
[DEBUG   ] Seconds remaining: 0.668332                                                                                                                                                                                                        
[DEBUG   ] _splay: 2019-01-30 11:54:02 _next_scheduled_fire_time: 2019-01-30 11:54:02 _next_fire_time: 2019-01-30 11:54:00                     
[DEBUG   ] {'_next_fire_time': datetime.datetime(2019, 1, 30, 11, 54), '_run': True, 'function': 'cmd.run', 'splay': 10, 'args': ['date >> /tmp/splay_test.log'], 'name': 'splay_test', 'when': ['11:54am'], 'enabled': True, 'jid_include': T
rue, '_next_scheduled_fire_time': datetime.datetime(2019, 1, 30, 11, 54, 2), '_splay': datetime.datetime(2019, 1, 30, 11, 54, 2), 'maxrunning': 1}
```

**Comment by rbthomp** [comment]
I remember having similar issues a ways back. Is this related to #50216 at all?

**Comment by sfozz** [comment]
Fixed in #50216
